control/state.py

```diff
@@ -163,159 +270,214 @@ def _remove_key(self, key: str):
             self.logger.error(f"Unable to remove key from omap: {ex}. Exiting!")
             raise
 
-    def add_bdev(self, bdev_name: str, val: str):
-        """Adds a bdev to the OMAP."""
-        key = self.BDEV_PREFIX + bdev_name
-        self._add_key(key, val)
+        if self.use_notify:  # Notify other gateways within the group of change
+            try:
+                notify = self.ioctx.notify(self.omap_name)
+            except Exception as ex:
+                self.logger.info(f"Failed to notify.")
 
-    def remove_bdev(self, bdev_name: str):
-        """Removes a bdev from the OMAP."""
-        key = self.BDEV_PREFIX + bdev_name
-        self._remove_key(key)
+    def delete_state(self):
+        """Deletes OMAP object contents."""
+        try:
+            with rados.WriteOpCtx() as write_op:
+                self.ioctx.clear_omap(write_op)
+                self.ioctx.operate_write_op(write_op, self.omap_name)
+                self.ioctx.set_omap(write_op, (self.OMAP_VERSION_KEY,),
+                                    (str(1),))
+                self.ioctx.operate_write_op(write_op, self.omap_name)
+                self.logger.info(f"Deleted OMAP contents.")
+        except Exception as ex:
+            self.logger.error(f"Error deleting OMAP contents: {ex}. Exiting!")
+            raise
 
-    def _restore_bdevs(self, omap_dict, callback):
-        """Restores a bdev from the OMAP."""
+    def register_watch(self, callback):
+        """Sets a watch on the OMAP object for changes."""
 
-        for (key, val) in omap_dict.items():
-            if key.startswith(self.BDEV_PREFIX):
-                # Get bdev_name from end of key
-                bdev_name = key.split("_", 1)[1]
-                req = json_format.Parse(val, pb2.create_bdev_req())
-                req.bdev_name = bdev_name
-                callback(req)
+        def _watcher_callback(notify_id, notifier_id, watch_id, data):
+            callback()
 
-    def add_namespace(self, subsystem_nqn: str, nsid: str, val: str):
-        """Adds a namespace to the OMAP."""
-        key = self.NAMESPACE_PREFIX + subsystem_nqn + "_" + nsid
-        self._add_key(key, val)
+        try:
+            self.use_notify = True
+            self.watch = self.ioctx.watch(self.omap_name, _watcher_callback)
```

Comment:
Define `watch` in `__init__` and set to `None`.
What if register_watch is called again? Should we check if `self.watch` is already set?